### ISSUE TITLE
INVALID (was: systemd-resolved makes up PTR records from a cached A record lookup)

### ISSUE THREAD

**Issue by wdoekes** [systemd/systemd#7110]
### Submission type

 - Bug report

### systemd version the issue has been seen with

systemd        232-21ubuntu5 amd64
on Ubuntu Zesty

### In case of bug report: Expected behaviour you didn't see

$ dig -t ptr 206.212.58.216.in-addr.arpa. @8.8.8.8 +short
ams16s21-in-f14.1e100.net.
ams16s21-in-f206.1e100.net.

### In case of bug report: Unexpected behaviour you saw

$ dig -t a google.com +short
216.58.212.206

$ dig -t ptr 206.212.58.216.in-addr.arpa. +short
google.com.

^-- this is wrong. systemd-resolved should not just make up PTR records; the correct PTR record would have been the two ams16 records as seen above

### Solution ###

Can you fix so systemd-resolve queries the nameserver for the PTR record instead of returning a cached record that is irrelevant?

Thanks :)

If this has been fixed already, I'll happily forward this to launchpad, with an appropriate patch if possible.

**Comment by floppym** [comment]
Seems to work for me on systemd-235.

```
% dig @127.0.0.53 -t a google.com

; <<>> DiG 9.11.2 <<>> @127.0.0.53 -t a google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 11785
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;google.com.                    IN      A

;; ANSWER SECTION:
google.com.             227     IN      A       216.58.192.174

;; Query time: 29 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: Mon Oct 16 17:15:43 EDT 2017
;; MSG SIZE  rcvd: 55

% dig @127.0.0.53 -x 216.58.192.174

; <<>> DiG 9.11.2 <<>> @127.0.0.53 -x 216.58.192.174
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 34155
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;174.192.58.216.in-addr.arpa.   IN      PTR

;; ANSWER SECTION:
174.192.58.216.in-addr.arpa. 86337 IN   PTR     ord36s02-in-f174.1e100.net.
174.192.58.216.in-addr.arpa. 86337 IN   PTR     ord36s02-in-f14.1e100.net.

;; Query time: 108 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: Mon Oct 16 17:16:09 EDT 2017
;; MSG SIZE  rcvd: 126
```

**Comment by wdoekes** [comment]
Thanks for checking!

So, I browsed the dns-related changes between v232 and master, looking for changes in e.g. calls to answer_add_addresses_ptr/dns_cache_put. Couldn't find anything. Nor did I find any intentional ptr synthesizing.

Next I tried to reproduce on my desktop, also with Zesty. Couldn't.

I'll have to go back to my laptop / reproducible environment and see what's going on. Perhaps systemd-resolved is not the culprit here. To be continued..

**Comment by wdoekes** [comment]
Sorry about this. Turns out it was my router :(

I'm going to go ahead and say that my router has received a new "feature" at some point in 2017, since it has 2017-firmware which *I* didn't install, but my cable company apparently did. And because I switched to systemd-resolved and ran into a Domains= bug, I jumped to the conclusion that systemd was to blame here as well. It was not.

For future reference: **Draytek Vigor2132FVn - 3.7.9.1 - (built: Apr 21 2017 17:29:16)** - makes up PTR records. For every A record that it has cached for the duration of TTL, it returns a PTR record of 60s.

Mea culpa. Closing as invalid.
